src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingClientProvider.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2023 Snowflake Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.snowflake.kafka.connector.internal.streaming;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.snowflake.kafka.connector.SnowflakeSinkConnectorConfig;
+import com.snowflake.kafka.connector.internal.KCLogger;
+import java.util.Map;
+import java.util.concurrent.locks.ReentrantLock;
+import net.snowflake.ingest.streaming.SnowflakeStreamingIngestClient;
+
+/**
+ * Singleton/factory that provides the streaming client(s). There should only be one provider, but
+ * it may provide multiple clients
+ */
```

Comment:
nit: Singleton and Factory are two different concepts. (Not an or)

```suggestion
/**
 * Factory that provides the streaming client(s). There should only be one provider, but
 * it may provide multiple clients if optimizations are disabled. (Look at ENABLE_STREAMING_CLIENT_OPTIMIZATION_CONFIG)
 */
```

Reply:
Thats true, the StreamingClientProvider is more of the factory that provides the StreamingClientProviderSingleton, since it can return any number of clients